uroot_test.go

```diff
@@ -0,0 +1,198 @@
+	c := exec.Command(filepath.Join(goroot, "bin/go"), "build", filepath.Join(gopath, "src/..."))
+	c.Env = append(b.env,
+		fmt.Sprintf("GOPATH=%s", gopath),
+		fmt.Sprintf("GOROOT=%s", goroot))
+	out, err := c.CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("could not build go source %v; output\n%s", err, out)
+	}
+	return nil
+}
+
+func TestUrootCmdline(t *testing.T) {
+	t.Parallel()
+
+	samplef, err := ioutil.TempFile("", "u-root-sample-")
+	if err != nil {
+		t.Fatal(err)
+	}
+	samplef.Close()
+	defer os.RemoveAll(samplef.Name())
+
+	for _, tt := range []struct {
+		name       string
+		env        []string
+		args       []string
+		err        error
+		validators []itest.ArchiveValidator
+	}{
+		{
+			name: "include one extra file",
+			args: []string{"-nocmd", "-files=/bin/bash"},
+			err:  nil,
+			validators: []itest.ArchiveValidator{
+				itest.HasFile{"bin/bash"},
+			},
+		},
+		{
+			name: "include multiple extra files",
+			args: []string{"-nocmd", "-files=/bin/bash", "-files=/bin/ls", fmt.Sprintf("-files=%s", samplef.Name())},
+			validators: []itest.ArchiveValidator{
+				itest.HasFile{"bin/bash"},
+				itest.HasFile{"bin/ls"},
+				itest.HasFile{samplef.Name()},
+			},
+		},
+		{
+			name: "include one extra file with rename",
+			args: []string{"-nocmd", "-files=/bin/bash:bin/bush"},
+			validators: []itest.ArchiveValidator{
+				itest.HasFile{"bin/bush"},
+			},
+		},
+		{
+			name: "hosted source mode",
+			args: append([]string{"-build=source", "-base=/dev/null", "-defaultsh=", "-initcmd="}, srcmds...),
+		},
+		{
+			name: "hosted bb mode",
+			args: append([]string{"-build=bb", "-base=/dev/null", "-defaultsh=", "-initcmd="}, twocmds...),
+		},
+		{
+			name: "AMD64 bb build",
+			env:  []string{"GOARCH=amd64"},
+			args: []string{"-build=bb"},
+		},
+		{
+			name: "AMD64 source build",
+			env:  []string{"GOARCH=amd64"},
+			args: []string{"-build=source"},
+			validators: []itest.ArchiveValidator{
+				buildSourceValidator{
+					goroot: "/go",
+					gopath: ".",
+					env:    []string{"GOARCH=amd64"},
+				},
+			},
+		},
+		{
+			name: "ARM7 bb build",
+			env:  []string{"GOARCH=arm", "GOARM=7"},
+			args: []string{"-build=bb"},
+		},
+		{
+			name: "ARM64 bb build",
+			env:  []string{"GOARCH=arm64"},
+			args: []string{"-build=bb"},
+		},
+		{
+			name: "Power 64bit bb build",
+			env:  []string{"GOARCH=ppc64le"},
+			args: []string{"-build=bb"},
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			f, sum1 := buildIt(t, tt.args, tt.env, tt.err)
+			defer os.RemoveAll(f.Name())
+
+			a, err := itest.ReadArchive(f.Name())
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			for _, v := range tt.validators {
+				if err := v.Validate(a); err != nil {
+					t.Errorf("validator failed: %v / archive:\n%s", err, a)
+				}
+			}
+
+			f2, sum2 := buildIt(t, tt.args, tt.env, tt.err)
+			defer os.RemoveAll(f2.Name())
+			if !bytes.Equal(sum1, sum2) {
+				t.Fatalf("not reproducible, hashes don't match")
```

Comment:
Can you also print the path to the two temp files and not delete them on failure?